File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid/Presenter/SharedView/SharedView_PopupControlDateTime.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.Graphics.Drawables;
using Android.OS;
using Android.Runtime;
using Android.Support.V7.App;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Class;
using BPMOPMobile.Droid.Class;
using BPMOPMobile.Droid.Core.Common;
using BPMOPMobile.Droid.Core.Controller;
using BPMOPMobile.Droid.Presenter.Fragment;
using Newtonsoft.Json.Linq;

namespace BPMOPMobile.Droid.Presenter.SharedView
{
    class SharedView_PopupControlDateTime : SharedView_PopupControlBase
    {
        private ControllerBase CTRLBase = new ControllerBase();
        public SharedView_PopupControlDateTime(LayoutInflater _inflater, AppCompatActivity _mainAct, CustomBaseFragment _fragment, string _fragmentTag, View _rootView) : base(_inflater, _mainAct, _fragment, _fragmentTag, _rootView)
        {

        }

        public override void InitializeValue_Master(ViewElement clickedElement)
        {
            base.InitializeValue_Master(clickedElement);
        }

        public override void InitializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JObject JObjectChild)
        {
            base.InitializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
        }

        public override void InitializeView()
        {
            base.InitializeView();
            try
            {
                #region Get View - Init Data
                View _viewPopupControl = _inflater.Inflate(Resource.Layout.PopupControl_DateTimePicker, null);
                DatePicker _datePicker = _viewPopupControl.FindViewById<DatePicker>(Resource.Id.dp_PopupControl_DateTimePicker);
                TimePicker _timePicker = _viewPopupControl.FindViewById<TimePicker>(Resource.Id.tp_PopupControl_DateTimePicker);
                ImageView _imgDelete = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_DateTimePicker_Delete);
                ImageView _imgToday = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_DateTimePicker_Today);
                ImageView _imgClose = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_DateTimePicker_Close);
                TextView _tvTitle = _viewPopupControl.FindViewById<TextView>(Resource.Id.tv_PopupControl_DateTimePicker_Title);
                LinearLayout _lnApply = _viewPopupControl.FindViewById<LinearLayout>(Resource.Id.ln_PopupControl_DateTimePicker_Clear);
                TextView _tvApply = _viewPopupControl.FindViewById<TextView>(Resource.Id.tv_PopupControl_DateTimePicker_Apply);

                _timePicker.SetIs24HourView(Java.Lang.Boolean.True);

                _tvApply.Text = CmmFunction.GetTitle("TEXT_APPLY", "Áp dụng");

                switch (base.flagView)
                {
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                    default:
                        {
                            DateTime _initDate = new DateTime();
                            try
                            {
                                _initDate = DateTime.Parse(elementParent.Value, new CultureInfo("en", false));
                                _datePicker.Init(_initDate.Year, _initDate.Month - 1, _initDate.Day, null);
                                _timePicker.Hour = _initDate.Hour;
                                _timePicker.Minute = _initDate.Minute;
                            }
                            catch
                            {
                                _initDate = DateTime.Now;
                            }

                            if (CmmVariable.SysConfig.LangCode.Equals(CmmDroidVariable.M_SysLangVN))
                                _tvTitle.Text = _initDate.ToString("dd/MM/yyyy");
                            else
                                _tvTitle.Text = _initDate.ToString("MM/dd/yyyy");


                            if (elementParent.Enable)
                            {
                                _imgDelete.Visibility = ViewStates.Visible;
                                _imgToday.Visibility = ViewStates.Visible;
                                _tvApply.Visibility = ViewStates.Visible;
                                _lnApply.Visibility = ViewStates.Visible;
                                _datePicker.Enabled = true;
                                _timePicker.Enabled = true;
                            }
                            else
                            {
                                _imgDelete.Visibility = ViewStates.Invisible;
                                _imgToday.Visibility = ViewStates.Invisible;
                                _tvApply.Visibility = ViewStates.Gone;
                                _lnApply.Visibility = ViewStates.Gone;
                                _datePicker.Enabled = false;
                                _timePicker.Enabled = false;
                            }

                            break;
                        }
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                        {
                            DateTime _initDate = new DateTime();
                            try
                            {
                                _initDate = DateTime.Parse(elementPopup.Value, new CultureInfo("en", false));
                                _datePicker.Init(_initDate.Year, _initDate.Month - 1, _initDate.Day, null);
                                _timePicker.Hour = _initDate.Hour;
                                _timePicker.Minute = _initDate.Minute;
                            }
                            catch
                            {
                                _initDate = DateTime.Now;
                            }

                            if (CmmVariable.SysConfig.LangCode.Equals(CmmDroidVariable.M_SysLangVN))
                                _tvTitle.Text = _initDate.ToString("dd/MM/yyyy");
                            else
                                _tvTitle.Text = _initDate.ToString("MM/dd/yyyy");

                            if (elementPopup.Enable)
                            {
                                _imgDelete.Visibility = ViewStates.Visible;
                                _imgToday.Visibility = ViewStates.Visible;
                                _tvApply.Visibility = ViewStates.Visible;
                                _lnApply.Visibility = ViewStates.Visible;
                                _datePicker.Enabled = true;
                                _timePicker.Enabled = true;
                            }
                            else
                            {
                                _imgDelete.Visibility = ViewStates.Invisible;
                                _imgToday.Visibility = ViewStates.Invisible;
                                _tvApply.Visibility = ViewStates.Gone;
                                _lnApply.Visibility = ViewStates.Gone;
                                _datePicker.Enabled = false;
                                _timePicker.Enabled = false;
                            }
                            break;
                        }
                }
                #endregion

                #region Show View                
                Dialog _dialogPopupControl = new Dialog(_rootView.Context);
                Window window = _dialogPopupControl.Window;
                _dialogPopupControl.RequestWindowFeature(1);
                _dialogPopupControl.SetCanceledOnTouchOutside(false);
                _dialogPopupControl.SetCancelable(true);
                window.SetSoftInputMode(SoftInput.AdjustPan); // Để Hide soft keyboard ko bị Resize
                window.SetGravity(GravityFlags.Center);
                var dm = _mainAct.Resources.DisplayMetrics;

                _dialogPopupControl.SetContentView(_viewPopupControl);
                _dialogPopupControl.Show();
                WindowManagerLayoutParams s = window.Attributes;
                s.Width = dm.WidthPixels;
                s.Height = WindowManagerLayoutParams.WrapContent;
                window.Attributes = s;
                window.SetBackgroundDrawable(new ColorDrawable(Color.Transparent));
                #endregion

                #region Event
                _imgClose.Click += (sender, e) =>
                {
                    _dialogPopupControl.Dismiss();
                };

                _imgToday.Click += (sender, e) =>
                {
                    DateTime _initDate = DateTime.Now;
                    _datePicker.Init(_initDate.Year, _initDate.Month - 1, _initDate.Day, null);
                    _timePicker.Hour = _initDate.Hour;
                    _timePicker.Minute = _initDate.Minute;
                };

                _imgDelete.Click += (sender, e) =>
                {
                    switch (base.flagView)
                    {
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForElement(elementParent, "");
                                _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, "");
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                    }
                };

                _lnApply.Click += delegate
                {
                    switch (base.flagView)
                    {
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                            {
                                DateTime _resultDate = new DateTime(_datePicker.Year, _datePicker.Month + 1, _datePicker.DayOfMonth, _timePicker.Hour, _timePicker.Minute, 0);
                                string _result = _resultDate.ToString("yyyy-MM-dd hh:mm:ss tt", CultureInfo.InvariantCulture);

                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForElement(elementParent, _result);
                                _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                _dialogPopupControl.Dismiss();

                                break;
                            }
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                            {
                                DateTime _resultDate = new DateTime(_datePicker.Year, _datePicker.Month + 1, _datePicker.DayOfMonth, _timePicker.Hour, _timePicker.Minute, 0);
                                string _result = _resultDate.ToString("yyyy-MM-dd hh:mm:ss tt", CultureInfo.InvariantCulture);

                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, _result);
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                    }
                };
                #endregion

            }
            catch (Exception ex)
            {
#if DEBUG
                CmmDroidFunction.WriteTrackingError(this.GetType().Name, "InitializeView", ex);
#endif
            }
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/shareview/SharedView_PopupControlDateTime.java
package com.vuthao.bpmop.shareview;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.DatePicker;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.TimePicker;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.gson.Gson;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.CalendarUltis;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.vars.VarsReceiver;
import com.vuthao.bpmop.core.Vars;
import com.vuthao.bpmop.detail.custom.DetailFunc;

import org.json.JSONObject;

import java.util.Calendar;

public class SharedView_PopupControlDateTime extends SharedView_PopupControlBase implements View.OnClickListener {
    private Dialog dialog;
    private View view;
    private DatePicker datePicker;
    private TimePicker timePicker;
    private ImageView imgDelete;
    private ImageView imgToday;
    private ImageView imgClose;
    private TextView tvTitle;
    private LinearLayout lnApply;
    private TextView tvApply;

    public SharedView_PopupControlDateTime(LayoutInflater inflater, Activity mainAct, String fragmentTag, View rootView) {
        super(inflater, mainAct, fragmentTag, rootView);
    }

    @Override
    public void initializeValue_Master(ViewElement viewElement) {
        super.initializeValue_Master(viewElement);
    }

    @Override
    public void initializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JSONObject JObjectChild) {
        super.initializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
    }

    @Override
    public void initializeView() {
        super.initializeView();

        view = inflater.inflate(R.layout.popup_control_datetime_picker, null);
        datePicker = view.findViewById(R.id.dp_PopupControl_DateTimePicker);
        timePicker = view.findViewById(R.id.tp_PopupControl_DateTimePicker);
        imgDelete = view.findViewById(R.id.img_PopupControl_DateTimePicker_Delete);
        imgToday = view.findViewById(R.id.img_PopupControl_DateTimePicker_Today);
        imgClose = view.findViewById(R.id.img_PopupControl_DateTimePicker_Close);
        tvTitle = view.findViewById(R.id.tv_PopupControl_DateTimePicker_Title);
        lnApply = view.findViewById(R.id.ln_PopupControl_DateTimePicker_Clear);
        tvApply = view.findViewById(R.id.tv_PopupControl_DateTimePicker_Apply);

        timePicker.setIs24HourView(Boolean.TRUE);
        tvApply.setText(Functions.share.getTitle("TEXT_APPLY", "Áp dụng"));
        CalendarUltis calendar = null;
        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                if (!Functions.isNullOrEmpty(elementParent.getValue())) {
                    calendar = new CalendarUltis(elementParent.getValue());
                    tvTitle.setText(DetailFunc.share.formatDateTimePopup(elementParent.getValue()));
                } else {
                    calendar = new CalendarUltis();
                    tvTitle.setText(Functions.share.formatLongToString(calendar.getTimeInMillis(), "dd/MM/yyyy"));
                }

                datePicker.init(calendar.YEAR, calendar.MONTH, calendar.DAY_OF_MONTH, null);
                timePicker.setHour(calendar.HOUR);
                timePicker.setMinute(calendar.MINUTE);

                if (elementParent.isEnable()) {
                    imgDelete.setVisibility(View.VISIBLE);
                    imgToday.setVisibility(View.VISIBLE);
                    tvApply.setVisibility(View.VISIBLE);
                    lnApply.setVisibility(View.VISIBLE);
                    datePicker.setEnabled(true);
                    timePicker.setEnabled(true);
                } else {
                    imgDelete.setVisibility(View.INVISIBLE);
                    imgToday.setVisibility(View.INVISIBLE);
                    tvApply.setVisibility(View.GONE);
                    lnApply.setVisibility(View.GONE);
                    datePicker.setEnabled(false);
                    timePicker.setEnabled(false);
                }
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                if (!Functions.isNullOrEmpty(elementPopup.getValue())) {
                    calendar = new CalendarUltis(elementPopup.getValue());
                    tvTitle.setText(DetailFunc.share.formatDateTimePopup(elementParent.getValue()));
                } else {
                    calendar = new CalendarUltis();
                    tvTitle.setText(Functions.share.formatLongToString(calendar.getTimeInMillis(), "dd/MM/yyyy"));
                }

                datePicker.init(calendar.YEAR, calendar.MONTH, calendar.DAY_OF_MONTH, null);
                timePicker.setHour(calendar.HOUR);
                timePicker.setMinute(calendar.MINUTE);

                tvTitle.setText(DetailFunc.share.formatDateTimePopup(elementParent.getValue()));

                if (elementPopup.isEnable()) {
                    imgDelete.setVisibility(View.VISIBLE);
                    imgToday.setVisibility(View.VISIBLE);
                    tvApply.setVisibility(View.VISIBLE);
                    lnApply.setVisibility(View.VISIBLE);
                    datePicker.setEnabled(true);
                    timePicker.setEnabled(true);
                } else {
                    imgDelete.setVisibility(View.INVISIBLE);
                    imgToday.setVisibility(View.INVISIBLE);
                    tvApply.setVisibility(View.GONE);
                    lnApply.setVisibility(View.GONE);
                    datePicker.setEnabled(false);
                    timePicker.setEnabled(false);
                }
                break;
            }
        }

        //region Dialog
        dialog = new Dialog(mainAct);
        Window window = dialog.getWindow();
        dialog.requestWindowFeature(1);
        dialog.setCanceledOnTouchOutside(false);
        dialog.setCancelable(true);
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        window.setGravity(Gravity.CENTER);
        DisplayMetrics displayMetrics = mainAct.getResources().getDisplayMetrics();
        dialog.setContentView(view);
        dialog.show();

        WindowManager.LayoutParams s = window.getAttributes();
        s.width = displayMetrics.widthPixels;
        s.height = WindowManager.LayoutParams.WRAP_CONTENT;
        window.setAttributes(s);
        window.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        //endregion

        imgToday.setOnClickListener(this);
        imgDelete.setOnClickListener(this);
        lnApply.setOnClickListener(this);
        imgClose.setOnClickListener(this);
    }

    private void today() {
        CalendarUltis c = new CalendarUltis();
        datePicker.init(c.YEAR, c.MONTH, c.DAY_OF_MONTH, null);
        timePicker.setHour(c.HOUR);
        timePicker.setMinute(c.MINUTE);
    }

    private void delete() {
        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATEFORM);
                intent.putExtra("element", new Gson().toJson(elementParent));
                intent.putExtra("newValue", "");
                BroadcastUtility.send(mainAct, intent);
                dialog.dismiss();
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATECHILDACTION);
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("newValue", "");
                BroadcastUtility.send(mainAct, intent);
                dialog.dismiss();
                break;
            }
        }
    }

    private void apply() {
        int day = datePicker.getDayOfMonth();
        int month = datePicker.getMonth();
        int year = datePicker.getYear();
        int hour = timePicker.getHour();
        int minute = timePicker.getMinute();

        Calendar calendar1 = Calendar.getInstance();
        calendar1.set(year, month, day, hour, minute);
        String result = Functions.share.formatLongToString(calendar1.getTimeInMillis(), "yyyy-MM-dd hh:mm:ss");

        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATEFORM);
                intent.putExtra("element", new Gson().toJson(elementParent));
                intent.putExtra("newValue", result);
                BroadcastUtility.send(mainAct, intent);
                dialog.dismiss();
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATECHILDACTION);
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("newValue", result);
                BroadcastUtility.send(mainAct, intent);
                dialog.dismiss();
                break;
            }
        }
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.img_PopupControl_DateTimePicker_Today: {
                today();
                break;
            }
            case R.id.img_PopupControl_DateTimePicker_Delete: {
                delete();
                break;
            }
            case R.id.ln_PopupControl_DateTimePicker_Clear: {
                apply();
                break;
            }
            case R.id.img_PopupControl_DateTimePicker_Close: {
                dialog.dismiss();
                break;
            }
        }
    }
}
